pkg/operator/controllers/image/image_controller_test.go

```diff
@@ -0,0 +1,176 @@
+			Features: arov1alpha1.FeaturesSpec{
+				ReconcileImageConfig: true,
+			},
+			Location: "eastus",
+		},
+	})
+)
+
+// Test reconcile function
+func TestImageConfigReconciler(t *testing.T) {
+	log := logrus.NewEntry(logrus.StandardLogger())
+	type test struct {
+		name       string
+		arocli     aroclient.Interface
+		configcli  configclient.Interface
+		wantConfig string
+	}
+
+	for _, tt := range []*test{
+		{
+			name: "Feature Flag disabled, no action",
+			arocli: arofake.NewSimpleClientset(&arov1alpha1.Cluster{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: arov1alpha1.SingletonClusterName,
+				},
+				Spec: arov1alpha1.ClusterSpec{
+					ACRDomain:     "arointsvc.azurecr.io",
+					AZEnvironment: "AzurePublicCloud",
+					Features: arov1alpha1.FeaturesSpec{
+						ReconcileImageConfig: false,
+					},
+					Location: "eastus",
+				},
+			}),
+			configcli: configfake.NewSimpleClientset(&configv1.Image{
+				ObjectMeta: imageConfigMetadata,
+				Spec: configv1.ImageSpec{
+					RegistrySources: configv1.RegistrySources{
+						AllowedRegistries: []string{
+							"quay.io",
+						},
+					},
+				},
+			}),
+			wantConfig: `{"metadata":{"name":"cluster","creationTimestamp":null},"spec":{"additionalTrustedCA":{"name":""},"registrySources":{"allowedRegistries":["quay.io"]}},"status":{}}`,
+		},
+		{
+			name:   "Image config registry source is empty, no action",
+			arocli: arocli,
+			configcli: configfake.NewSimpleClientset(&configv1.Image{
+				ObjectMeta: imageConfigMetadata,
+			}),
+			wantConfig: `{"metadata":{"name":"cluster","creationTimestamp":null},"spec":{"additionalTrustedCA":{"name":""},"registrySources":{}},"status":{}}`,
+		},
+		{
+			name:   "allowedRegistries exists, function should add images",
+			arocli: arocli,
+			configcli: configfake.NewSimpleClientset(&configv1.Image{
+				ObjectMeta: imageConfigMetadata,
+				Spec: configv1.ImageSpec{
+					RegistrySources: configv1.RegistrySources{
+						AllowedRegistries: []string{
+							"quay.io",
+						},
+					},
+				},
+			}),
+			wantConfig: `{"metadata":{"name":"cluster","creationTimestamp":null},"spec":{"additionalTrustedCA":{"name":""},"registrySources":{"allowedRegistries":["quay.io","arointsvc.azurecr.io","arosvc.eastus.data.azurecr.io"]}},"status":{}}`,
+		},
+		{
+			name:   "blockedRegistries exists, function should delete images",
+			arocli: arocli,
+			configcli: configfake.NewSimpleClientset(&configv1.Image{
+				ObjectMeta: imageConfigMetadata,
+				Spec: configv1.ImageSpec{
+					RegistrySources: configv1.RegistrySources{
+						BlockedRegistries: []string{
+							"quay.io", "arointsvc.azurecr.io", "arosvc.eastus.data.azurecr.io",
+						},
+					},
+				},
+			}),
+			wantConfig: `{"metadata":{"name":"cluster","creationTimestamp":null},"spec":{"additionalTrustedCA":{"name":""},"registrySources":{"blockedRegistries":["quay.io"]}},"status":{}}`,
+		},
+		{
+			name: "Gov Cloud, function should remove appropriate images",
```

Comment:
Another idea for a test case would be if `instance.Spec.AZEnvironment` is unset / missing.